### PR TITLE
add gluon model summary

### PR DESCRIPTION
**Pull request by szha**
## Description ##
add hook/pre-hook to gluon block and model summary

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] add `register_forward_hook`, `register_forward_pre_hook`, `apply`
- [x] add model summary

### Notes ###
- Currently, only the obvious container classes such as Sequential/HybridSequential are skipped. For models that are purely container with child blocks registered through `setattr` are not skipped because they cannot be easily identified.

### CODE REVIEW

File: python/mxnet/gluon/block.py [hunk 1]
```diff
@@ -355,14 +357,68 @@ def load_params(self, filename, ctx=None, allow_missing=False,
...
+        Parameters
+        ----------
+        hook : callable
+            The forward hook function of form `hook(block, input) -> None`.
+
+        Returns
+        -------
+        :class:`mxnet.gluon.utils.RemovableHandle`
+        """
+        handle = RemovableHandle(self._forward_pre_hooks)
```
[the remark this answers is not in the document]
Comment (szha) [reply python/mxnet/gluon/block.py]: OK

**Comment by piiswrong** [comment]
Need warning for hybridize. otherwise LGTM

**Comment by szha** [comment]
Thanks. I'm still working on a general model summary method, and this PR will need another round of review on that.

**Comment by szha** [comment]
## Some examples

### Resnet50 V1
```python
    net = gluon.model_zoo.vision.resnet50_v1()
    net.initialize()
    net.summary(mx.nd.ones((32, 3, 224, 224)))
```
Results:
```bash
--------------------------------------------------------------------------------
        Layer (type)                                Output Shape         Param #
================================================================================
               Input                           (32, 3, 224, 224)               0
            Conv2D-1                          (32, 64, 112, 112)            9408
         BatchNorm-2                          (32, 64, 112, 112)             256
        Activation-3                          (32, 64, 112, 112)               0
         MaxPool2D-4                            (32, 64, 56, 56)               0
            Conv2D-5                            (32, 64, 56, 56)            4160
         BatchNorm-6                            (32, 64, 56, 56)             256
        Activation-7                            (32, 64, 56, 56)               0
            Conv2D-8                            (32, 64, 56, 56)           36864
         BatchNorm-9                            (32, 64, 56, 56)             256
       Activation-10                            (32, 64, 56, 56)               0
           Conv2D-11                           (32, 256, 56, 56)           16640
        BatchNorm-12                           (32, 256, 56, 56)            1024
           Conv2D-13                           (32, 256, 56, 56)           16384
        BatchNorm-14                           (32, 256, 56, 56)            1024
     BottleneckV1-15                           (32, 256, 56, 56)               0
           Conv2D-16                            (32, 64, 56, 56)           16448
        BatchNorm-17                            (32, 64, 56, 56)             256
       Activation-18                            (32, 64, 56, 56)               0
           Conv2D-19                            (32, 64, 56, 56)           36864
        BatchNorm-20                            (32, 64, 56, 56)             256
       Activation-21                            (32, 64, 56, 56)               0
           Conv2D-22                           (32, 256, 56, 56)           16640
        BatchNorm-23                           (32, 256, 56, 56)            1024
     BottleneckV1-24                           (32, 256, 56, 56)               0
           Conv2D-25                            (32, 64, 56, 56)           16448
        BatchNorm-26                            (32, 64, 56, 56)             256
       Activation-27                            (32, 64, 56, 56)               0
           Conv2D-28                            (32, 64, 56, 56)           36864
        BatchNorm-29                            (32, 64, 56, 56)             256
       Activation-30                            (32, 64, 56, 56)               0
           Conv2D-31                           (32, 256, 56, 56)           16640
        BatchNorm-32                           (32, 256, 56, 56)            1024
     BottleneckV1-33                           (32, 256, 56, 56)               0
           Conv2D-34                           (32, 128, 28, 28)           32896
        BatchNorm-35                           (32, 128, 28, 28)             512
       Activation-36                           (32, 128, 28, 28)               0
           Conv2D-37                           (32, 128, 28, 28)          147456
        BatchNorm-38                           (32, 128, 28, 28)             512
       Activation-39                           (32, 128, 28, 28)               0
           Conv2D-40                           (32, 512, 28, 28)           66048
        BatchNorm-41                           (32, 512, 28, 28)            2048
           Conv2D-42                           (32, 512, 28, 28)          131072
        BatchNorm-43                           (32, 512, 28, 28)            2048
     BottleneckV1-44                           (32, 512, 28, 28)               0
           Conv2D-45                           (32, 128, 28, 28)           65664
        BatchNorm-46                           (32, 128, 28, 28)             512
       Activation-47                           (32, 128, 28, 28)               0
           Conv2D-48                           (32, 128, 28, 28)          147456
        BatchNorm-49                           (32, 128, 28, 28)             512
       Activation-50                           (32, 128, 28, 28)               0
           Conv2D-51                           (32, 512, 28, 28)           66048
        BatchNorm-52                           (32, 512, 28, 28)            2048
     BottleneckV1-53                           (32, 512, 28, 28)               0
           Conv2D-54                           (32, 128, 28, 28)           65664
        BatchNorm-55                           (32, 128, 28, 28)             512
       Activation-56                           (32, 128, 28, 28)               0
           Conv2D-57                           (32, 128, 28, 28)          147456
        BatchNorm-58                           (32, 128, 28, 28)             512
       Activation-59                           (32, 128, 28, 28)               0
           Conv2D-60                           (32, 512, 28, 28)           66048
        BatchNorm-61                           (32, 512, 28, 28)            2048
     BottleneckV1-62                           (32, 512, 28, 28)               0
           Conv2D-63                           (32, 128, 28, 28)           65664
        BatchNorm-64                           (32, 128, 28, 28)             512
       Activation-65                           (32, 128, 28, 28)               0
           Conv2D-66                           (32, 128, 28, 28)          147456
        BatchNorm-67                           (32, 128, 28, 28)             512
       Activation-68                           (32, 128, 28, 28)               0
           Conv2D-69                           (32, 512, 28, 28)           66048
        BatchNorm-70                           (32, 512, 28, 28)            2048
     BottleneckV1-71                           (32, 512, 28, 28)               0
           Conv2D-72                           (32, 256, 14, 14)          131328
        BatchNorm-73                           (32, 256, 14, 14)            1024
       Activation-74                           (32, 256, 14, 14)               0
           Conv2D-75                           (32, 256, 14, 14)          589824
        BatchNorm-76                           (32, 256, 14, 14)            1024
       Activation-77                           (32, 256, 14, 14)               0
           Conv2D-78                          (32, 1024, 14, 14)          263168
        BatchNorm-79                          (32, 1024, 14, 14)            4096
           Conv2D-80                          (32, 1024, 14, 14)          524288
        BatchNorm-81                          (32, 1024, 14, 14)            4096
     BottleneckV1-82                          (32, 1024, 14, 14)               0
           Conv2D-83                           (32, 256, 14, 14)          262400
        BatchNorm-84                           (32, 256, 14, 14)            1024
       Activation-85                           (32, 256, 14, 14)               0
           Conv2D-86                           (32, 256, 14, 14)          589824
        BatchNorm-87                           (32, 256, 14, 14)            1024
       Activation-88                           (32, 256, 14, 14)               0
           Conv2D-89                          (32, 1024, 14, 14)          263168
        BatchNorm-90                          (32, 1024, 14, 14)            4096
     BottleneckV1-91                          (32, 1024, 14, 14)               0
           Conv2D-92                           (32, 256, 14, 14)          262400
        BatchNorm-93                           (32, 256, 14, 14)            1024
       Activation-94                           (32, 256, 14, 14)               0
           Conv2D-95                           (32, 256, 14, 14)          589824
        BatchNorm-96                           (32, 256, 14, 14)            1024
       Activation-97                           (32, 256, 14, 14)               0
           Conv2D-98                          (32, 1024, 14, 14)          263168
        BatchNorm-99                          (32, 1024, 14, 14)            4096
    BottleneckV1-100                          (32, 1024, 14, 14)               0
          Conv2D-101                           (32, 256, 14, 14)          262400
       BatchNorm-102                           (32, 256, 14, 14)            1024
      Activation-103                           (32, 256, 14, 14)               0
          Conv2D-104                           (32, 256, 14, 14)          589824
       BatchNorm-105                           (32, 256, 14, 14)            1024
      Activation-106                           (32, 256, 14, 14)               0
          Conv2D-107                          (32, 1024, 14, 14)          263168
       BatchNorm-108                          (32, 1024, 14, 14)            4096
    BottleneckV1-109                          (32, 1024, 14, 14)               0
          Conv2D-110                           (32, 256, 14, 14)          262400
       BatchNorm-111                           (32, 256, 14, 14)            1024
      Activation-112                           (32, 256, 14, 14)               0
          Conv2D-113                           (32, 256, 14, 14)          589824
       BatchNorm-114                           (32, 256, 14, 14)            1024
      Activation-115                           (32, 256, 14, 14)               0
          Conv2D-116                          (32, 1024, 14, 14)          263168
       BatchNorm-117                          (32, 1024, 14, 14)            4096
    BottleneckV1-118                          (32, 1024, 14, 14)               0
          Conv2D-119                           (32, 256, 14, 14)          262400
       BatchNorm-120                           (32, 256, 14, 14)            1024
      Activation-121                           (32, 256, 14, 14)               0
          Conv2D-122                           (32, 256, 14, 14)          589824
       BatchNorm-123                           (32, 256, 14, 14)            1024
      Activation-124                           (32, 256, 14, 14)               0
          Conv2D-125                          (32, 1024, 14, 14)          263168
       BatchNorm-126                          (32, 1024, 14, 14)            4096
    BottleneckV1-127                          (32, 1024, 14, 14)               0
          Conv2D-128                             (32, 512, 7, 7)          524800
       BatchNorm-129                             (32, 512, 7, 7)            2048
      Activation-130                             (32, 512, 7, 7)               0
          Conv2D-131                             (32, 512, 7, 7)         2359296
       BatchNorm-132                             (32, 512, 7, 7)            2048
      Activation-133                             (32, 512, 7, 7)               0
          Conv2D-134                            (32, 2048, 7, 7)         1050624
       BatchNorm-135                            (32, 2048, 7, 7)            8192
          Conv2D-136                            (32, 2048, 7, 7)         2097152
       BatchNorm-137                            (32, 2048, 7, 7)            8192
    BottleneckV1-138                            (32, 2048, 7, 7)               0
          Conv2D-139                             (32, 512, 7, 7)         1049088
       BatchNorm-140                             (32, 512, 7, 7)            2048
      Activation-141                             (32, 512, 7, 7)               0
          Conv2D-142                             (32, 512, 7, 7)         2359296
       BatchNorm-143                             (32, 512, 7, 7)            2048
      Activation-144                             (32, 512, 7, 7)               0
          Conv2D-145                            (32, 2048, 7, 7)         1050624
       BatchNorm-146                            (32, 2048, 7, 7)            8192
    BottleneckV1-147                            (32, 2048, 7, 7)               0
          Conv2D-148                             (32, 512, 7, 7)         1049088
       BatchNorm-149                             (32, 512, 7, 7)            2048
      Activation-150                             (32, 512, 7, 7)               0
          Conv2D-151                             (32, 512, 7, 7)         2359296
       BatchNorm-152                             (32, 512, 7, 7)            2048
      Activation-153                             (32, 512, 7, 7)               0
          Conv2D-154                            (32, 2048, 7, 7)         1050624
       BatchNorm-155                            (32, 2048, 7, 7)            8192
    BottleneckV1-156                            (32, 2048, 7, 7)               0
 GlobalAvgPool2D-157                            (32, 2048, 1, 1)               0
           Dense-158                                  (32, 1000)         2049000
        ResNetV1-159                                  (32, 1000)               0
================================================================================
Total params: 25629032
Trainable params: 25575912
Non-trainable params: 53120
--------------------------------------------------------------------------------
```

### Single input/output RNN
```python
    net2 = nn.Sequential()
    with net2.name_scope():
        net2.add(nn.Embedding(10, 20))
        net2.add(gluon.rnn.LSTM(30))
        net2.add(nn.Dense(40, flatten=False))
    net2.initialize()
    net2.summary(mx.nd.ones((80, 32)))
```
Results:
```bash
--------------------------------------------------------------------------------
        Layer (type)                                Output Shape         Param #
================================================================================
               Input                                    (80, 32)               0
         Embedding-1                                (80, 32, 20)             200
              LSTM-2                                (80, 32, 30)            6240
             Dense-3                                (80, 32, 40)            1240
================================================================================
Total params: 7680
Trainable params: 7680
Non-trainable params: 0
--------------------------------------------------------------------------------
```

### Multi-input/output RNN
```python
    net3 = gluon.rnn.LSTM(30)
    net3.initialize()
    begin_state = net3.begin_state(32)
    net3.summary(mx.nd.ones((80, 32, 5)), begin_state)
```
Results:
```bash
--------------------------------------------------------------------------------
        Layer (type)                                Output Shape         Param #
================================================================================
               Input     (80, 32, 5), [(1, 32, 30), (1, 32, 30)]               0
              LSTM-1    (80, 32, 30), [(1, 32, 30), (1, 32, 30)]            4440
================================================================================
Total params: 4440
Trainable params: 4440
Non-trainable params: 0
--------------------------------------------------------------------------------
```

**Comment by safrooze** [comment]
@szha I think this change causes a regression reported in [issue #11772](https://github.com/apache/incubator-mxnet/issues/11772).

**Pull request merged**